File: legacy-typescript/src/source/prsd.ts
export type PrsdBlock = {
  index: number;
  code: string;
};

export function extractPrescribeBlocks(text: string): PrsdBlock[] {
  const lines = text.split(/\r?\n/);
  const blocks: PrsdBlock[] = [];
  let inBlock = false;
  let buffer: string[] = [];
  let blockIndex = 0;

  for (const line of lines) {
    if (!inBlock) {
      if (line.trim() === ":::prescribe") {
        inBlock = true;
        buffer = [];
      }
      continue;
    }

    if (line.trim() === ":::") {
      blocks.push({ index: blockIndex++, code: buffer.join("\n") });
      inBlock = false;
      buffer = [];
      continue;
    }

    buffer.push(line);
  }

  if (inBlock) {
    blocks.push({ index: blockIndex++, code: buffer.join("\n") });
  }

  return blocks;
}

File: csharp/Prescribe.Core/Source/Prsd.cs
namespace Prescribe.Core.Source;

public sealed record PrsdBlock(int Index, string Code);

public static class Prsd
{
    public static IReadOnlyList<PrsdBlock> ExtractPrescribeBlocks(string text)
    {
        var lines = text.Split(new[] { "\r\n", "\n" }, StringSplitOptions.None);
        var blocks = new List<PrsdBlock>();
        var inBlock = false;
        var buffer = new List<string>();
        var blockIndex = 0;

        foreach (var line in lines)
        {
            if (!inBlock)
            {
                if (line.Trim() == ":::prescribe")
                {
                    inBlock = true;
                    buffer.Clear();
                }
                continue;
            }

            if (line.Trim() == ":::")
            {
                blocks.Add(new PrsdBlock(blockIndex++, string.Join("\n", buffer)));
                inBlock = false;
                buffer.Clear();
                continue;
            }

            buffer.Add(line);
        }

        if (inBlock)
        {
            blocks.Add(new PrsdBlock(blockIndex++, string.Join("\n", buffer)));
        }

        return blocks;
    }
}
